garrysmod/gamemodes/sandbox/gamemode/cl_hints.lua

```diff
@@ -1,62 +1,80 @@
 
+local cl_showhints = CreateConVar( "cl_showhints", "1",
+	bit.bor( FCVAR_ARCHIVE, FCVAR_USERINFO ), "Show hints called thru SANDBOX:AddHint", 0, 1 )
 
-CreateClientConVar( "cl_showhints", "1", true, false )
+-- A list of hints we've already done so we don't repeat ourselves
+GM.ProcessedHints = {}
 
--- A list of hints we've already done so we don't repeat ourselves`
-local ProcessedHints = {}
+local function ThrowHint( GM, name, length, sound )
 
---
--- Throw's a Hint to the screen
---
-local function ThrowHint( name )
-
-	local show = GetConVarNumber( "cl_showhints" )
-	if ( show == 0 ) then return end
-
-	if ( engine.IsPlayingDemo() ) then return end
+	if ( !cl_showhints:GetBool() ) then return end
 
 	local text = language.GetPhrase( "Hint_" .. name )
-
-	local s, e, group = string.find( text, "%%([^%%]+)%%" )
-	while ( s ) do
-		local key = input.LookupBinding( group )
-		if ( !key ) then key = "<NOT BOUND>" end
-
-		text = string.gsub( text, "%%([^%%]+)%%", "'" .. key:upper() .. "'" )
-		s, e, group = string.find( text, "%%([^%%]+)%%" )
+	local startpos, endpos, bind = text:find( "%%([^%%]+)%%" )
+
+	if ( startpos ~= nil ) then
+		local buffer = {}
+		local bufferlen = 1
+		local nextstartpos = 1
+		
+		repeat
+			local key = input.LookupBinding( bind )
+
+			if ( key ~= nil ) then
+				key = key:upper()
+			else
+				key = bind .. " (command not bound to a key)"
+			end
+
+			-- Add the unaltered string + the translated key to the string buffer
+			buffer[ bufferlen ] = text:sub( nextstartpos, startpos - 1 )
+			buffer[ bufferlen + 1 ] = key
+			bufferlen = bufferlen + 2
+
+			nextstartpos = endpos + 1
+			startpos, endpos, bind = text:find( "%%([^%%]+)%%", nextstartpos )
+		until ( startpos == nil )
+		
+		text = table.concat( buffer, "", 1, bufferlen - 1 )
 	end
 
-	GAMEMODE:AddNotify( text, NOTIFY_HINT, 20 )
+	if ( length == nil ) then length = 20 end
+	GM:AddNotify( text, NOTIFY_HINT, length )
 
-	surface.PlaySound( "ambient/water/drip" .. math.random( 1, 4 ) .. ".wav" )
+	if ( sound == nil ) then sound = "ambient/water/drip" .. math.random( 1, 4 ) .. ".wav" end
+	surface.PlaySound( sound )
 
 end
 
+function GM:AddHint( name, delay, length, sound )
 
---
--- Adds a hint to the queue
---
-function GM:AddHint( name, delay )
+	if ( self.ProcessedHints[ name ] ) then return end
 
-	if ( ProcessedHints[ name ] ) then return end
+	if ( !engine.IsPlayingDemo() ) then
+		timer.Create( "GMOD_HintSystem_" .. name, delay, 1, function()

```

Comment:
What is the point of adding `GMOD_` prefix to the timer's ID? Do we not know what game we are?

Reply:
Timer identifiers are global. Hook identifiers are global. ConCommands/ConVars are global. Many other built-in systems (Lua or engine based) share common identifiers - adding the prefix is just a formality to avoid conflict.